elasticdl/python/allreduce/tensorflow_optimizer.py

```diff
@@ -0,0 +1,671 @@
+      op:
+        The reduction operation to use when combining gradients across
+        different ranks.
+      gradient_predivide_factor:
+        If op == Average, gradient_predivide_factor splits the averaging
+        before and after the sum. Gradients are scaled by
+        1.0 / gradient_predivide_factor before the sum and
+        gradient_predivide_factor / size after the sum.
+      average_aggregated_gradients:
+        Whether to average the aggregated gradients that have been accumulated
+        over multiple mini-batches. If true divides gradients updates by
+        backward_passes_per_step. Only applicable for
+        backward_passes_per_step > 1.
+      num_groups:
+        Number of groups to assign gradient allreduce ops to for explicit
+        grouping. Defaults to no explicit groups.
+    fixed_global_batch_size:
+        Whether to keep the global batch size is fixed even though the worker
+        number is changing during elastic execution.
+    global_batch_count_per_step:
+        Total batch number per step of all the workers.
+    """
+
+    # *ElasticDL Update*: If `fixed_global_batch_size` == False,
+    # just fallback to the native horovod DistributedOptimizer.
+    if not fixed_global_batch_size:
+        return hvd.DistributedOptimizer(
+            optimizer=optimizer,
+            name=name,
+            use_locking=use_locking,
+            device_dense=device_dense,
+            device_sparse=device_sparse,
+            compression=compression,
+            sparse_as_dense=sparse_as_dense,
+            backward_passes_per_step=backward_passes_per_step,
+            op=op,
+            gradient_predivide_factor=gradient_predivide_factor,
+            average_aggregated_gradients=average_aggregated_gradients,
+            num_groups=num_groups,
+        )
+
+    if gradient_predivide_factor != 1.0:
+        if hvd.rocm_built():
+            raise ValueError(
+                "gradient_predivide_factor not supported yet with ROCm"
+            )
+        if op != hvd.Average:
+            raise ValueError(
+                "gradient_predivide_factor not supported with op != Average"
+            )
+
+    if op == hvd.Adasum and average_aggregated_gradients:
+        raise ValueError(
+            "Adasum does not support average_aggregated_gradients == True"
+        )
+
+    if isinstance(optimizer, _LegacyOptimizer):
+        if op == hvd.Adasum:
+            return hvd._DistributedAdasumOptimizer(
+                optimizer,
+                name,
+                use_locking,
+                device_dense,
+                device_sparse,
+                compression,
+                backward_passes_per_step,
+            )
+
+        # If global_batch_count_per_step is None from the parameter,
+        # try to get the value from environment variable.
+        global_batch_count_per_step = complement_value_from_env_if_none(
+            global_batch_count_per_step, "global_batch_count_per_step", int
```

Comment:
Now,  we can set "WORKER_NUM" in env of workers. 
https://github.com/sql-machine-learning/elasticdl/blob/d7d081b43ae05525d83d73aa6d430ececfcf5ba5/elasticdl/python/master/pod_manager.py#L140-L142

Can we get the `global_batch_count_per_step` by following codes?
```python
worker_num = int(os.getenv("WORKER_NUM", 0))
global_batch_count_per_step = (
    backward_passes_per_step * worker_num
    if global_batch_count_per_step is None
    else global_batch_count_per_step
)
```

Reply:
Good idea, updated.